specs/granite/header_accumulator.md

```diff
@@ -0,0 +1,302 @@
+
+## Overview
+
+The Granite hardfork introduces a new phase of block execution, header hash accumulation. After fork activation,
+every `HEADER_BATCH_SIZE` blocks, the execution layer must append a binary merkle tree of depth
+`HEADER_BATCH_TREE_DEPTH` containing the previous `HEADER_BATCH_SIZE` block hashes (exclusive of the current block) to
+the header accumulation merkle tree of depth `ACCUMULATOR_TREE_DEPTH` in the previous header accumulation block.
+
+## Timestamp Activation
+
+Granite, like other OP Stack network upgrades, is activated at a timestamp. Changes to the L2 Block execution rules
+outlined in this specification are applied when the `L2 Timestamp >= activation time`.
+
+## Rationale
+
+After the activation of the [interop hardfork](../interop/overview.md), the computational complexity of proving the
+OP Stack's state transition will increase. The current plan for proving the inclusion of remote logs within chains
+that exist in the active dependency set is to include a final "consolidation step" to resolve the dependencies of
+each chain, which currently could require deriving and executing many L2 blocks in the context of the verifiable
+environment that the proving system is executing within.
+
+This feature seeks to reduce the complexity of the consolidation step significantly, through allowing the consolidation
+program to produce only a single L2 block at a height that includes _all receipts_ on a single chain that must be
+validated to resolve its dependents. With this, the computational complexity of resolving a dependency is drastically
+reduced, only requiring two merkle proofs as well as `n` merkle patricia trie proofs to verify inclusion within the
+receipt roots within the headers commited to in the accumulator tree.
+
+This feature also allows for very efficient historical state lookups for other situations that are periphery to this
+optimization for interop. For example, in the fault proof program, this feature removes the need for a commit-reveal
+walkback when retrieving data within the historical chain. Instead, we would only need to provide small inclusion
+proofs for a constant-time lookup of any data in the historical state accumulator.
+
+## Constants & Definitions
+
+| Term                      | Description                                                                                                                                |
+| ------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------ |
+| `ACCUMULATOR_TREE_DEPTH`  | `27`                                                                                                                                       |
+| `HEADER_BATCH_TREE_DEPTH` | `5`                                                                                                                                        |
```

Comment:
This comes out to a 32 leaf nodes, and I think that means a 32x performance optimization, is that correct? As the header concerns now happen only once per `HEADER_BATCH_SIZE`?

Do any behaviors change if we find that this needs to be increased further? Like, if we had 64 per batch, would we start to put pressure on other parts of the system in negative ways?

Reply:
Yes, that is the optimization. 32x less data stored in `extraData`, and it is only appended to the header.

I would advise not making `HEADER_BATCH_SIZE` > 256, since most execution layer clients drop state > 256 blocks old out of memory. With `HEADER_BATCH_SIZE <= 256`, the lookups for the header hashes should be very quick, since we can directly consult in-memory state at tip.